### PR TITLE
[onert] Introduce loss map for training

### PR DESCRIPTION
**Pull request by jyoungyun**
This commit introduces loss map for accessing training loss value from nnfw internal api.

ONE-DCO-1.0-Signed-off-by: Jiyoung Yun <jy910.yun@samsung.com>

Draft: #11035

### CODE REVIEW

File: runtime/onert/core/include/ir/train/TrainableGraph.h [hunk 1]
```diff
@@ -101,6 +101,7 @@ class TrainableGraph : public IGraph
   void changeDerivativeShape(const OperandIndex &ind, const ir::Shape &new_shape);
   void addInput(const OperandIndex &ind, const std::string &name = "");
   void addOutput(const OperandIndex &ind, const std::string &name = "");
+  void addLoss(const IOIndex &io_ind, const OperandIndex &ind);
```
Comment (ragmani) [inline runtime/onert/core/include/ir/train/TrainableGraph.h:104]: How about changing the name of the parameter to make it easier to know the index for which operand in Loss?

Comment (jyoungyun) [reply runtime/onert/core/include/ir/train/TrainableGraph.h:104]:
```suggestion
  void addLoss(const IOIndex &out_ind, const OperandIndex &loss_ind);
```

How about this?

Comment (ragmani) [reply runtime/onert/core/include/ir/train/TrainableGraph.h]: I mean the name of operand. How about this?
```suggestion
  void addLoss(const IOIndex &true_io_ind, const OperandIndex &true_ind);
```

Comment (jyoungyun) [reply runtime/onert/core/include/ir/train/TrainableGraph.h]: I fixed this code to `pred_io_ind` and `loss_ind`.

The loss map has a combination of `pred_io_ind` and the loss output index of loss op for `pred_io_ind` output.

**Comment by jyoungyun** [comment]
@nnfw-bot test onert-x64-release

**Review by ragmani — DISMISSED** [review]
LGTM

File: runtime/onert/core/include/ir/train/TrainableGraph.h [hunk 2]
```diff
@@ -101,6 +101,7 @@ class TrainableGraph : public IGraph
   void changeDerivativeShape(const OperandIndex &ind, const ir::Shape &new_shape);
   void addInput(const OperandIndex &ind, const std::string &name = "");
   void addOutput(const OperandIndex &ind, const std::string &name = "");
+  void addLoss(const IOIndex &pred_io_ind, const OperandIndex &loss_ind);
```
Comment (glistening) [inline runtime/onert/core/include/ir/train/TrainableGraph.h:104]: - What does `addLoss`?
- What is `pred_io_ind` and `loss_ind`?

 I am trying to figure out by reading`LossInsertionPass` from #11035.

Comment (jyoungyun) [reply runtime/onert/core/include/ir/train/TrainableGraph.h:104]: The `addLoss` inserts a pair information about `pred_io_ind` and `loss_ind`.
`LossInsertionPass` inserts new Loss operation for calculating loss value using prediction value and expected value. User can invoke the `nnfw_train_get_loss` function to obtain loss value about each output index. Currently there is no way to get the index of the newly added Loss Op for output index. This map has information about the index of Loss Op for a given output index.

Comment (glistening) [reply runtime/onert/core/include/ir/train/TrainableGraph.h:104]: @jyoungyun Thank you for explanation.

Here is my understanding.

- `addLoss` adds a key-value pair into unordered_map (`pred_io_ind`<sup>IOIndex</sup> → `loss_ind` <sup>OperandIndex</sup>).
- `pred_io_ind` is simply the output index. (Currently and simple case, it is 0.)
- `loss_ind` is the _output_ operand of `Loss` operation.
- `addLoss` will be called as many times as the number of model output.

Comment (glistening) [reply runtime/onert/core/include/ir/train/TrainableGraph.h:104]: (optional)

I would like to make `addLoss` similar to `addInput` and `addOutput`.

```cc
void addInput(const OperandIndex &ind, const std::string &name = "");
void addOutput(const OperandIndex &ind, const std::string &name = "");
```

`addInput` and `addOutput` inform `TrainableGraph`
- 1st argument (OperandIndex) is input (or output)
- 2nd argument gives additional information about the OperandIndex


```suggestion
  void addLoss(const OperandIndex &loss_ind, const IOIndex &pred_io_ind);
```

It will inform `TrainableGraph`
- 1st argument (OperandIndex) is loss.
- 2nd argument gives additional information that IOIndex is associated with the operand.

Comment (jyoungyun) [reply runtime/onert/core/include/ir/train/TrainableGraph.h]: I applied your suggestion. Please review this PR again.

**Review by glistening — APPROVED** [review]
LGTM

**Review by ragmani — APPROVED** [review]
